test/unit/math/test_ad.hpp

```diff
@@ -1632,6 +1715,25 @@ void check_return_type(const ReturnType& ret, const Type& x) {
   }
 }
 
+/**
+ * Given the input and outputs are `std::vector` types, check that if
+ * the input value type is a var<matrix> then the output value_type should be
+ * a `var<matrix>` as well. Similarly if the input value_type is
+ * `matrix<var>`, the output value type should be `matrix<var>`.
+ *
+ * @tparam ReturnType The result of applying `x` to a function `f()`
+ * @tparam Type the type of the input to the function `f()`.
+ * @param ret output of function
+ * @param x input of function
+ */
+template <typename ReturnType, typename Type,
+          require_std_vector_t<ReturnType>* = nullptr,
+          require_std_vector_vt<is_matrix, Type>* = nullptr>
+void check_return_type(const ReturnType& ret, const Type& x) {
+  if (ret.size() > 0 && x.size() > 0)
+    check_return_type(ret[0], x[0]);
```

Comment:
I will not be offended if you want me to take a closer look at `check_return_type`.

I think the logic we want is:

1. If there are only `var_value<double>` s on the input, then there should only be `var_value<double>` s on the output
2. If there are `var_value<not double>` s  on the input, then there should be no `var_value<double>` s on the output

I think this would need an extra template program `var_value_t` to extract the `var_value` from a generic input type and I am too lazy to write it today.

Reply:
I think the logic is fine here unless I'm missing something. This is a specialization for `std::vector<T>` that just checks that the inner type for the input/output is correct. If that's what this does then I think it's fine